### ISSUE TITLE
Use environment variables on rsyslog configuration (imkafka/omkafka - confparam)

### ISSUE THREAD

**Issue by Karl90** [rsyslog/rsyslog#3887]
Dear Rsyslog Team,

We're attempting to pass certain values from environment variables to rsyslog configuration, for the "confparam" parameter from the modules imkafka/omkafka:

https://www.rsyslog.com/doc/master/configuration/modules/omkafka.html#confparam

However, we haven't been able to do it properly, note that the array requires a list of params surrounded by double quotes.

### Expected behavior

Read environment variables on the rsyslog configuration, 
specifically in the context of imkafka/omkafka modules: trying to pass values on the "confparam" array.

### Actual behavior

The environment variables from the confparam are not being picked up, maybe a bad config? or it is actually not possible?

### Steps to reproduce the behavior

Try to pass an environment variable to the confparam array, for instance:

[rsyslog@3b6ad7a869e0 ~]$ echo $KPASS
jUiAmnfJWkKV/Hp8GGQRi42hE
[rsyslog@3b6ad7a869e0 ~]$ grep KPASS /etc/rsyslog.d/kafka_publisher.conf
        confParam=["linger.ms=100", "security.protocol=SASL_SSL", "sasl.mechanisms=PLAIN", "sasl.username=Publisher", "sasl.password=`echo $KPASS`"]

When running rsyslog in debug mode, we can actually see that the env variable is not taken, but the literal string:

5441.718788800:kafka-publish-ruleset queue:Reg/w0: omkafka.c: omkafka: setting custom configuration parameter: sasl.password:`echo $KPASS`
...
6240.412972800:kafka-publish-ruleset queue:Reg/w0: errmsg.c: Called LogMsg, msg: omkafka: action will suspended due to kafka error -187: Local: All broker connections are down
6240.412984000:kafka-publish-ruleset queue:Reg/w0: operatingstate.c: osf: MSG omkafka: action will suspended due to kafka error -187: Local: All broker connections are down: 
6240.413004400:kafka-publish-ruleset queue:Reg/w0: iminternal.c: signaling new internal message via SIGTTOU: 'omkafka: action will suspended due to kafka error -187: Local: All broker connections are down [v8.1910.0 try https://www.rsyslog.com/e/2422 ]'
6240.413004400:kafka-publish-ruleset queue:Reg/w0: errmsg.c: Called LogMsg, msg: omkafka: action will suspended due to kafka error -169: Local: Authentication failure

since the value is not taken, the Kafka authentication fails, as a result publishing/subscription is not possible.
Note that this would be actually helpful on containerised environment where sensitive keys can be stored as secrets and then passed to rsyslog through environment variables/files.

### Environment
- rsyslog version: v8.1910.0
- platform: Centos 7 (containerised)
- for configuration questions/issues, include rsyslog.conf and included config files:

Note that the confparam accepts an array of parameters (strings), any librdkafka property can be set through this array in this way {"param=value"}, i.e.:

       confParam=["compression.codec=snappy",
                  "socket.timeout.ms=5",
                  "socket.keepalive.enable=true"]

However, we haven't been able to set the value reading an environment value, in my example:

                 "sasl.password=`echo $KPASS`"

Is there a way to do this properly?
As showed, in this case the literal string (`echo $KPASS`) is taken as value.

So far, the only way found is to include these values in clear text on the configuration file and it is not desired, since some of them are sensible.

I can see that rsyslog supports this feature

https://github.com/rsyslog/rsyslog/pull/2419

But we haven't been able to do it for this case, may I request your help to check this?

Thanks in advance,

Karl.

**Comment by rgerhards** [comment]
This is tricky as you need to change the array parameter. I admit I think it currently is not support, but have no time now to dig deep. You may want to try exporting the full setting as env var
```
export KPASS_PARAM="sasl.password=$KPASS`"
```
and then do
```
confParam=["linger.ms=100", "security.protocol=SASL_SSL", "sasl.mechanisms=PLAIN", "sasl.username=Publisher", `echo $KPASS_PARAM`]
```
But I am not even sure if the grammar supports it. Please let me know the outcome.

**Comment by Karl90** [comment]
Hi @rgerhards ,

Thanks a lot for your suggestion, it was almost right :-)

For reference, I've tried with these env vars:

```
[rsyslog@9f3336ab67bb /]# echo $KPASS_PARAM1
"sasl.password=jfrnjfnwejfnr"
[rsyslog@9f3336ab67bb /]# echo $KPASS_PARAM2
sasl.password=jfrnjfnwejfnr
```

In theory the KPASS_PARAM1 should be the appropriate setting, but actually throws an error:

Oct  4 20:59:00 9f3336ab67bb rsyslogd: error setting custom configuration parameter '"sasl.password=jfrnjfnwejfnr"': No such configuration property: ""sasl.password" [v8.1910.0 try https://www.rsyslog.com/e/1000 ]

(IDK why an extra quote is added)

However, when using the KPASS_PARAM2, it worked!

config:

`        
confParam=["security.protocol=SASL_SSL", "sasl.mechanisms=PLAIN", "sasl.username=username", `echo $KPASS_PARAM2`]
`

from debug:
```
4113.841423600:kafka-publish_ruleset queue:Reg/w0: omkafka.c: omkafka: setting custom configuration parameter: sasl.password:jfrnjfnwejfnr
...
4113.863794800:kafka-publish_ruleset queue:Reg/w0: omkafka.c: omkafka: writeKafka kafka outqueue length: 1, callbacks called 0
4113.863835800:kafka-publish_ruleset queue:Reg/w0: omkafka.c: omkafka: writeKafka returned 0
```

So, it seems that for the array, we don't need to include the surrounded quotes - as we actually do for the literal values.
As a result, we can consider this as misconfiguration and close this issue,

Thanks a lot for the hint, it did help!

Regards,

Karl.

**Comment by rgerhards** [comment]
thx for the confirmation. Just for the records: I made a typo in my posting:
```
export KPASS_PARAM="sasl.password=$KPASS`"
```
actually the backtick should not have been present. So the correct form is:
```
export KPASS_PARAM="sasl.password=$KPASS`"
```
I'll fix this now in the old posting. This was probably the reason it did not work initially. If you do the export like I intended, the double quote characters become not part of the string, and that is exactly how it should be (like you discovered).
